File: examples/governance_static_go/domain/assign_role.go
package domain

import (
	"time"
	"fmt"
)

type AssignRole struct {
	StakeholderId string `json:"stakeholder_id"`
	Role string `json:"role"`
}

func (c AssignRole) CommandName() string { return "AssignRole" }

func (c AssignRole) Execute(repo StakeholderRepository) (*Stakeholder, *AssignedRole, error) {
	existing, err := repo.Find(c.StakeholderId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("Stakeholder not found: %s", c.StakeholderId)
	}
	existing.Role = c.Role
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := AssignedRole{
		AggregateID: existing.ID,
		StakeholderId: c.StakeholderId,
		Role: c.Role,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}

File: examples/governance/identity_domain/lib/identity_domain/stakeholder/commands/assign_role.rb
module IdentityDomain
  class Stakeholder
    module Commands
      class AssignRole
        include Hecks::Command
        emits "AssignedRole"

        attr_reader :stakeholder_id, :role

        def initialize(stakeholder_id: nil, role: nil)
          @stakeholder_id = stakeholder_id
          @role = role
        end

        def call
          existing = repository.find(stakeholder_id)
          if existing
            Stakeholder.new(
              id: existing.id,
              name: existing.name,
              email: existing.email,
              team: existing.team,
              status: existing.status,
              role: role
            )
          else
            raise Hecks::Error, "Stakeholder not found: #{stakeholder_id}"
          end
        end
      end
    end
  end
end
